### ISSUE TITLE
Kindle Customisation stopped working

### ISSUE THREAD

**Issue by dunxd** [mikespub-org/seblucas-cops#36]
I was testing out some template changes with bootstrap5 for kindle, and noticed that the customisation page stopped setting the cookie and refreshing the page and selecting a new template (or style).

I rolled back commits till it started working again. It looks like it stopped working with https://github.com/mikespub-org/seblucas-cops/tree/26c9a44672d729454a0e8294ff6d7eb80c35de65

Last working commit is 5cc6e9e89cc9080eac41e50e4bc8eedd9e73061f :-(

I guess there must have been some dependancy of twig that uses javascript that doesn't work with the Kindle browser.

**Comment by mikespub** [comment]
I'm assuming you at least tried the 'twigged' template at some point? Otherwise I don't see how Kindle could be influenced by it...

Anyway, assuming you did try it, was this with server-side rendering or client-side rendering? It could be that the Kindle browser doesn't like the twig.js, but it shouldn't even be loading that since Kindle is supposed to be using server-side rendering.

So the only other piece I see is the change in the util.js file where I added the initiateTwig() function. Normally that shouldn't be a problem even if twig.js isn't loaded, but you might try to wrap this with something like:

```
if (typeof Twig !== 'undefined') {
    ...
}
```

Of course, it could be something completely unrelated to Twig, like the fact that jquery was updated to 3.7.1 in the last week or so and included in the latest docker images...

**Comment by dunxd** [comment]
I did try the twigged template - it works fine in a regular browser as far
as I've tried it.

I noticed the issue after creating a new version of bootstrap5 that doesn't
use any semantic html, as I think this might be what is preventing the
pages from loading when that template is set.

However, I didn't get to try it, as changing options on the customize page
doesn't do anything. Previously, when an option changed a cookie was set
accordingly, and in the case of template and style the page refreshed to
take on the changes. Neither of those things happen after the commit I
referred to.

It is very hard to troubleshoot since the Kindle doesn't show any errors or
even show page source. I can't change the template cookie on the kindle
without rolling back to a previous commit. I have confirmed that this is
independent of the template being used - default has the same issue.

The cookie code is all client side - the onchange events call
`updatecCookie()` which is in util.js. It is dependant on jQuery so I
suspect it is related to an update with that.

jQuery is loaded by composer right? I'm not confident I am testing
properly. Do I need to do anything between git checkouts to make sure all
the code is appropriately loaded?

**Comment by dunxd** [comment]
Commenting out the ititiateTwig() function in util.js made no difference to the behaviour.

**Comment by dunxd** [comment]
If I run `composer update` after switching commits will that ensure the correct versions of any dependancies are loaded and incorrect ones are removed?

**Comment by mikespub** [comment]
At least it's good to know changing util.js wouldn't help, thanks.

Regarding the checkouts, you should actually run
```
composer install -o
```
to (re-)install the versions that are defined in composer.lock for that checkout, and not composer update because that would actually update them :-)

For instance, jquery was updated from 3.7.0 to 3.7.1 about the same time I added twig templates in https://github.com/mikespub-org/seblucas-cops/commit/58a3d6c0dd65ff83574cfd7024957e7b902da243

**Comment by dunxd** [comment]
Unfortunately, on my production instance which is still using COPS 1.4.5 and for which the customize page works with Kindle experimental browser, the version of jQuery is also 3.7.1.

Using the Chrome development tools, I went through all the Sources loaded by the page in both versions. As far as I can see they are all on the same versions.

I also tested with my Home Assistant COPS container, which uses the ZIP release files, and rewrites the entire folder every update so composer shouldn't be an issue.  I get the same thing with recent releases:

* 2.0.1 - same issue
* 1.5.4 - same issue
* 1.5.0 - same issue
* 1.4.5 - working
* 1.4.3 - working

Ugh!

**Comment by dunxd** [comment]
Ok - I copied the util.js file from the 1.4.5 version over the top of the one in 1.5.0, and things worked.

So I checked the changes between the versions and commented them out, until I got it working again.  It does seem to be the initiateTwig function that is causing the issue. There are some other changes (lines 127 - 129 added, and the new way to update the title around line 325) but these don't cause the issue.

There must be something in the syntax of the initiateTwig function that the Kindle javascript engine doesn't like.

**Comment by dunxd** [comment]
I commented out the initiateTwig function in release 2.0.1 and the customize works ok on the Kindle.

**Comment by dunxd** [comment]
Wrapping the initiateTwig function in the if block you suggested does not fix it. I think it must be a syntax issue.

**Comment by dunxd** [comment]
I moved the initiateTwig() function block to templates/twigged/scripts/cops.js and customise is working ok on the Kindle.  

However, what is weird is that it also seems to be working ok if I switch to the twigged template and then change templates again.  I did some testing with the twigged template after making the above change, and essential functions such as browsing and downloading book files works on the Kindle.

**Comment by mikespub** [comment]
Very strange indeed. I'm not sure I understand the reason, especially since I remember trying to put the initializeTwig() where you put it once and it failed to work then - maybe simply due to browser caching...

Since I didn't care too much then, I was happy it worked fine in util.js, but now I guess I'll need revisit the issue ;-)

Glad you found a way to make it work though, even if we'll never understand why Kindle chokes on it...

**Comment by dunxd** [comment]
> maybe simply due to browser caching...

I hope that doesn't have anything to do with why things seemed to start working...

No, I did a _lot_ of switching versions and testing in the last 24 hours that makes me pretty certain I narrowed the issue down.

The kindle browser is horrible to work with. I need to make a Kindle specific template that has really large interface and a greyscale colour scheme that is good enough that we can set it as an override template for Kindle devices and never have to think about it again!
